packages/@react-aria/color/src/useColorAreaGradient.ts

```diff
@@ -10,205 +10,89 @@
-          dir = xChannel === 'red';
-          background = generateRGB_B(orientation, dir, zValue);
-          break;
-        }
-        case 'hue': {
-          dir = xChannel !== 'saturation';
-          if (isHSL) {
-            background = generateHSL_H(orientation, dir, zValue);
-          } else {
-            background = generateHSB_H(orientation, dir, zValue);
-          }
-          break;
-        }
-        case 'saturation': {
-          dir = xChannel === 'hue';
-          if (isHSL) {
-            background = generateHSL_S(orientation, dir, alphaValue);
-          } else {
-            background = generateHSB_S(orientation, dir, alphaValue);
-          }
-          break;
-        }
-        case 'brightness': {
-          dir = xChannel === 'hue';
-          background = generateHSB_B(orientation, dir, alphaValue);
-          break;
+
+    switch (state.value.getColorSpace()) {
+      case 'rgb': {
+        let rgb = parseColor('rgb(0, 0, 0)');
+        colorAreaStyles = {
+          background: [
+            // The screen blend mode multiplies the inverse of each channel, e.g. 1 - (1 - a) * (1 - b).
+            // Create a layer for each channel, with the other channels as 0. After blending, this should
+            // result in the gradients being combined channel by channel.
+            `linear-gradient(to ${end}, ${rgb.withChannelValue(xChannel, 0)}, ${rgb.withChannelValue(xChannel, 255)})`,
+            `linear-gradient(to top, ${rgb.withChannelValue(yChannel, 0)}, ${rgb.withChannelValue(yChannel, 255)})`,
+            rgb.withChannelValue(zChannel, zValue)
+          ].join(','),
+          backgroundBlendMode: 'screen'
+        };
+        break;
+      }
+      case 'hsl': {
+        let channels = state.value.getColorChannels();
+        let value = parseColor('hsl(0, 100%, 50%)').withChannelValue(zChannel, zValue);
+
+        let bg = channels
+          .filter(c => c !== zChannel)
+          .map(c => `linear-gradient(to ${c === xChannel ? end : 'top'}, ${hslChannels[c](value)})`)
+          .reverse();
+        if (zChannel === 'hue') {
+          bg.push(value.toString('css'));
         }
-        case 'lightness': {
-          dir = xChannel === 'hue';
-          background = generateHSL_L(orientation, dir, zValue);
-          break;
+    
+        colorAreaStyles = {
+          background: bg.join(', ')
+        };
+        break;
+      }
+      case 'hsb': {
+        let channels = state.value.getColorChannels();
+        let value = parseColor('hsb(0, 100%, 100%)').withChannelValue(zChannel, zValue);
+
+        let bg = channels
+          .filter(c => c !== zChannel)
+          .map(c => `linear-gradient(to ${c === xChannel ? end : 'top'}, ${hsbChannels[c](value)})`)
+          .reverse();
+        if (zChannel === 'hue') {
+          bg.push(value.toString('css'));
         }
+    
+        colorAreaStyles = {
+          background: bg.join(', ')
+        };
+        break;
       }
     }
```

Comment:
This refactors the way the gradients are generated to be both simpler and more correct. For RGB, we use `background-blend-mode` to combine the channel values rather than a transparency mask. This produces a more correct result than before, where the pixels at a given x/y position are closer to the actual value you get when you click. Some differences may appear in Chromatic for this.

For HSL and HSB, transparency is still used for saturation and lightness/brightness. But we had some bugs especially in HSB that are now fixed. For example, in the hue/saturation configuration, when lowering the brightness, the result shown in the thumb is much grayer than shown in the gradient. Same in the hue/brightness configuration with the saturation lowered.

<img width="248" alt="image" src="https://github.com/adobe/react-spectrum/assets/19409/1fa8ea5e-2d07-420c-9749-03b81b66fcdc"> 
<img width="257" alt="image" src="https://github.com/adobe/react-spectrum/assets/19409/52d308fb-d78b-4e6e-b57d-6d064fbfb7fd">

These are both now fixed.

<img width="253" alt="image" src="https://github.com/adobe/react-spectrum/assets/19409/11eb9631-ef01-4220-b6eb-d5b5791487d3">
<img width="253" alt="image" src="https://github.com/adobe/react-spectrum/assets/19409/0d8b4dbf-230d-4409-92a5-35355d5f0da0">

Reply:
Ran chromatic, it didn't appear to pick up on these https://www.chromatic.com/build?appId=5f0dd5ad2b5fc10022a2e320&number=737